anvio/synteny.py

```diff
@@ -0,0 +1,303 @@
+#!/usr/bin/env python
+# -*- coding: utf-8
+# pylint: disable=line-too-long
+
+"""Classes to work with ngrams of contig functions.
+
+These are classes to deconstruct loci into ngrams. They are used
+to analyze conserved genes and synteny structures across loci.
+"""
+
+import pandas as pd
+
+from collections import Counter
+
+import anvio
+import anvio.dbops as dbops
+import anvio.utils as utils
+import anvio.tables as t
+import anvio.terminal as terminal
+import anvio.filesnpaths as filesnpaths
+import anvio.genomedescriptions as genomedescriptions
+
+from anvio.errors import ConfigError
+
+
+__author__ = "Developers of anvi'o (see AUTHORS.txt)"
+__copyright__ = "Copyleft 2015-2018, the Meren Lab (http://merenlab.org/)"
+__credits__ = []
+__license__ = "GPL 3.0"
+__version__ = anvio.__version__
+__maintainer__ = "Matthew Schechter"
+__email__ = "mschechter@uchicago.edu"
+
+
+class NGram(object):
+    """class for counting NGrams
+
+    anvi-analyze-synteny is designed to work with a group of similar loci, where each locus is a contig (which
+    can lie in any number of contigs dbs)
+
+    Parameters
+    ==========
+    args : argparse.Namespace
+        For examples, arguments accepted by anvi-analyze-syntenty
+
+    skip_sanity_check : bool, False
+        If True, sanity_check will not be called.
+
+    Notes
+    =====
+    - Currently the design assumes that each locus is a contig. In the future we have plans to expand this to
+      compare genomes, not just loci, to each other. If that behavior is desired in the current design, each
+      genome should be a single contig.
+    """
+
+    def __init__(self, args, run=terminal.Run(), progress=terminal.Progress(), skip_sanity_check=False):
+        """Parses arguments and run sanity_check"""
+
+        self.args = args
+        self.run = run
+        self.progress = progress
+
+        # Parse arguments
+        A = lambda x: args.__dict__[x] if x in args.__dict__ else None
+        self.external_genomes = A('external_genomes')
+        self.annotation_source = A('annotation_source')
+        self.window_range = A('window_range')
```

Comment:
```
self.window_range = A('window_range') or 500
```

May be better. `anvio/__init__.py` has a default for this, but it is not the only way people can use this class. if I get an instance directly from the module (and not through `anvi-analyze-synteny` program) I wouldn't know what is the programmer suggested default for this.

Unfortunately we are not always careful about setting proper defaults for class variables throughout the codebase, but I wanted to point it out here anyway.

Reply:
Sorry Meren this concept was a little over my head but I think I understand the idea of default values for instances. I tried to fix this with this:

I added this to `synteny.py`:

```
 self.window_range = A('ngram_window_range') or "2:3"
```

and this to `anvio/__init__.py`:

```
'ngram-window-range': (
            ['--ngram-window-range'],
            {'default': "2:3",
             'metavar': "NGRAM_WINDOW_RANGE",
             'type': str,
             'required': False,

```